process/smartContract/process.go

```diff
@@ -421,27 +429,117 @@ func (sc *scProcessor) ProcessIfError(
 		return err
 	}
 
-	if !check.IfNil(acntSnd) {
-		err = acntSnd.AddToBalance(tx.GetValue())
+	if check.IfNil(acntSnd) {
+		moveBalanceCost := sc.economicsFee.ComputeFee(tx)
+		consumedFee.Sub(consumedFee, moveBalanceCost)
+	}
+
+	err = sc.addBackTxValues(acntSnd, scrIfError, tx)
+	if err != nil {
+		return err
+	}
+
+	err = sc.scrForwarder.AddIntermediateTransactions([]data.TransactionHandler{scrIfError})
+	if err != nil {
+		return err
+	}
+
+	err = sc.processForRelayerWhenError(tx, returnMessage)
+	if err != nil {
+		return err
+	}
+
+	sc.txFeeHandler.ProcessTransactionFee(consumedFee, big.NewInt(0), txHash)
+
+	return nil
+}
+
+func (sc *scProcessor) processForRelayerWhenError(
+	originalTx data.TransactionHandler,
+	returnMessage []byte,
+) error {
+	relayedSCR, isRelayedTx := isRelayedTx(originalTx)
+	if !isRelayedTx {
+		return nil
+	}
+
+	relayerAcnt, err := sc.getAccountFromAddress(relayedSCR.RelayerAddr)
+	if err != nil {
+		return err
+	}
+
+	if !check.IfNil(relayerAcnt) {
+		err := relayerAcnt.AddToBalance(relayedSCR.RelayedValue)
 		if err != nil {
 			return err
 		}
 
-		err = sc.accounts.SaveAccount(acntSnd)
+		err = sc.accounts.SaveAccount(relayerAcnt)
 		if err != nil {
 			log.Debug("error saving account")
+			return err
 		}
-	} else {
-		moveBalanceCost := sc.economicsFee.ComputeFee(tx)
-		consumedFee.Sub(consumedFee, moveBalanceCost)
 	}
 
-	err = sc.scrForwarder.AddIntermediateTransactions(scrIfError)
+	scrForRelayer := &smartContractResult.SmartContractResult{
+		Nonce:          originalTx.GetNonce(),
+		Value:          relayedSCR.RelayedValue,
+		RcvAddr:        relayedSCR.RelayerAddr,
+		SndAddr:        relayedSCR.RcvAddr,
+		OriginalTxHash: relayedSCR.OriginalTxHash,
+		ReturnMessage:  returnMessage,
+	}
+
+	err = sc.scrForwarder.AddIntermediateTransactions([]data.TransactionHandler{scrForRelayer})
 	if err != nil {
 		return err
 	}
 
-	sc.txFeeHandler.ProcessTransactionFee(consumedFee, big.NewInt(0), txHash)
+	return nil
+}
+
+func isRelayedTx(tx data.TransactionHandler) (*smartContractResult.SmartContractResult, bool) {
+	relayedSCR, ok := tx.(*smartContractResult.SmartContractResult)
+	if !ok {
+		return nil, false
+	}
+
+	if len(relayedSCR.RelayerAddr) == len(relayedSCR.SndAddr) &&
```

Comment:
Perhaps add a comment to explain the reason behind this condition?

Reply:
done